fea-rs/src/parse.rs

```diff
@@ -1,29 +1,79 @@
+pub use context::{IncludeStatement, ParseContext};
 pub use lexer::TokenSet;
 pub use parser::Parser;
-pub use source::{FileId, Source};
+pub use source::{Source, SourceLoadError, SourceResolver};
 pub use tree::ParseTree;
 
-pub(crate) use source::{SourceList, SourceMap};
+pub(crate) use source::{FileId, SourceList, SourceMap};
 
-use crate::GlyphMap;
+use crate::{Diagnostic, GlyphMap, Node};
 
-/// Attempt to parse a feature file at a given path, including its imports.
+/// Attempt to parse a feature file from disk, including its imports.
+///
+/// In general, you should not need to use this method directly; instead use one
+/// of the methods in the [`compile`][crate::compile] module.
+///
+/// The `project_root` argument is optional, and is intended for the case of
+/// handling UFO files, where imports are resolved relative to the root directory,
+/// and not the feature file itself.
+///
+/// The `glyph_map`, if provided, is used to disambiguate between certain tokens
+/// that are allowed in FEA syntax but which are also legal glyph names. If it
+/// is absent, and these names are encountered, we will report an error.
+///
+/// If you are compiling from memory, or otherwise want to handle loading files
+/// and resolving imports, you can use [`parse_root`] instead.
 pub fn parse_root_file(
     path: impl Into<PathBuf>,
     glyph_map: Option<&GlyphMap>,
     project_root: Option<PathBuf>,
-) -> Result<ParseContext, HardError> {
-    ParseContext::parse_from_root(path.into(), glyph_map, project_root)
+) -> Result<ParseContext, SourceLoadError> {
+    let path = path.into();
+    let project_root =
+        project_root.unwrap_or_else(|| path.parent().map(PathBuf::from).unwrap_or_default());
+    let resolver = source::FileSystemResolver::new(project_root);
+    parse_root(path.into_os_string(), glyph_map, resolver)
+}
+
+/// Entry point for parsing.
+///
+/// This method provides full control over how sources are located and include
+/// statements are resolved, by allowing you to provide a custom [`SourceResolver`].
+///
+/// The `path` argument is identifies the root source; it will be resolved against
+/// the provided `resolver`.
+///
+/// The `glyph_map`, if provided, is used to disambiguate between certain tokens
+/// that are allowed in FEA syntax but which are also legal glyph names. If you
+/// are not compiling the parse results, you can omit it.
+pub fn parse_root(
+    path: OsString,
+    glyph_map: Option<&GlyphMap>,
+    resolver: impl SourceResolver + 'static,
+) -> Result<ParseContext, SourceLoadError> {
+    ParseContext::parse_from_root(path.into(), glyph_map, resolver)
+}
+
+/// Convenience method to parse a block of FEA from memory.
+///
+/// This is useful for things like testing or syntax highlighting of a single file,
+/// but it cannot handle imports, or handle ambiguous glyph names.
+pub fn parse_string(text: impl Into<Arc<str>>) -> (Node, Vec<Diagnostic>) {
```

Comment:
I'm under the impression we're single threaded so ... why Arc not Rc?

Reply:
The boring answer is that the `Source` type was already using `Arc`, and I stuck with that. (I've since changed it to `Rc<str>`)